### ISSUE TITLE
NullPointerException: Attempt to invoke virtual method 'int com.google.android.gms.cast.MediaStatus.getLoadingItemId()' on a null object reference

### ISSUE THREAD

**Issue by Nimrodda** [googlecast/CastCompanionLibrary-android#205]
I'm getting the following exception everytime a video is finished and I'm in VideoCastControllerActivity.
any ideas?
```
java.lang.NullPointerException: Attempt to invoke virtual method 'int com.google.android.gms.cast.MediaStatus.getLoadingItemId()' on a null object reference
	at com.google.android.libraries.cast.companionlibrary.cast.player.VideoCastControllerFragment.updatePlayerStatus(VideoCastControllerFragment.java:469)
	at com.google.android.libraries.cast.companionlibrary.cast.player.VideoCastControllerFragment.access$1000(VideoCastControllerFragment.java:75)
	at com.google.android.libraries.cast.companionlibrary.cast.player.VideoCastControllerFragment$MyCastConsumer.onRemoteMediaPlayerStatusUpdated(VideoCastControllerFragment.java:259)
	at com.google.android.libraries.cast.companionlibrary.cast.VideoCastManager.onRemoteMediaPlayerStatusUpdated(VideoCastManager.java:2106)
	at com.google.android.libraries.cast.companionlibrary.cast.VideoCastManager.access$200(VideoCastManager.java:138)
	at com.google.android.libraries.cast.companionlibrary.cast.VideoCastManager$22.onStatusUpdated(VideoCastManager.java:1793)
	at com.google.android.gms.cast.RemoteMediaPlayer.onStatusUpdated(Unknown Source)
	at com.google.android.gms.cast.RemoteMediaPlayer.zza(Unknown Source)
	at com.google.android.gms.cast.RemoteMediaPlayer$1.onStatusUpdated(Unknown Source)
	at com.google.android.gms.cast.internal.zzm.zza(Unknown Source)
	at com.google.android.gms.cast.internal.zzm.zzbB(Unknown Source)
	at com.google.android.gms.cast.RemoteMediaPlayer.onMessageReceived(Unknown Source)
	at com.google.android.gms.cast.internal.zze$zzb$4.run(Unknown Source)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:135)
	at android.app.ActivityThread.main(ActivityThread.java:5254)
	at java.lang.reflect.Method.invoke(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:372)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
```

**Comment by naddaf** [comment]
I am not able to reproduce this. Can you try CastVideos-android and outline the steps that would result in this?

**Comment by Nimrodda** [comment]
I tried with CastVideos-android and it doesn't happen there.
I'm wondering what am I doing wrong.
When user clicks on a thumbnail, I get the `Video` object associated with that item in the `RecyclerView` and then basically start casting like so:

```
MediaInfo mediaInfo = buildMediaInfo(video.getTitle(), video.getLength(), video.getRenditions().get(renditionIndex).getUrl(), video.getThumbnail().getUrl());
				Bundle mediaWrapper = Utils.mediaInfoToBundle(mediaInfo);
				castManager.startVideoCastControllerActivity(mContext, mediaWrapper, startTime, true);
```

and `buildMediaInfo()` method:
```
private static MediaInfo buildMediaInfo(String title, long duration, String url, String bigImageUrl) {
		MediaMetadata movieMetadata = new MediaMetadata(MediaMetadata.MEDIA_TYPE_MOVIE);

		movieMetadata.putString(MediaMetadata.KEY_TITLE, title);
		movieMetadata.addImage(new WebImage(Uri.parse(bigImageUrl)));
		movieMetadata.addImage(new WebImage(Uri.parse(bigImageUrl)));

		return new MediaInfo.Builder(url)
				.setStreamType(MediaInfo.STREAM_TYPE_BUFFERED)
				.setContentType("video/mp4")
				.setMetadata(movieMetadata)
				.setStreamDuration(duration)
				.build();
	}
```

Am I missing something?
Thanks a lot for helping!

**Comment by Nimrodda** [comment]
I created an app from scratch only for the purpose of integrating CastCompanion and I get the exact same error. Please note that I followed the PDF documentation to do that, perhaps it's out-dated?

```
MediaMetadata mediaMetadata = new MediaMetadata();
				MediaInfo mediaInfo = new MediaInfo.Builder(URL)
						.setContentType("video/mp4")
						.setStreamType(MediaInfo.STREAM_TYPE_BUFFERED)
						.setMetadata(mediaMetadata).build();


				Bundle mediaWrapper = Utils.mediaInfoToBundle(mediaInfo);
				mCastManager.startVideoCastControllerActivity(v.getContext(), mediaWrapper, 0, true);
```

It seems identical to what you're doing in CastVideos.

**Comment by naddaf** [comment]
Since Queuing apis were introduced, I have moved away from that approach
and all my items are now part of a queue (might be a queue of one item).
Could you share your simplified sample so I can fix the issue? I do like to
support the older approach as well so that would be very helpful. If you
can't share that with me here, please send it to me via email, thanks.

**Comment by Nimrodda** [comment]
I've uploaded the sample project to Github and added you as contributor. Thanks for helping!
https://github.com/Nimrodda/ChromecastDemo

**Comment by Nimrodda** [comment]
In regards to the Queuing API, is that the new way to cast a video you referred to?

```
MediaQueueItem mediaQueueItem = new MediaQueueItem.Builder(mediaInfo).setAutoplay(true).setPreloadTime(20).build();
MediaQueueItem[] queueItems = new MediaQueueItem[] { mediaQueueItem };
try {
	mCastManager.queueLoad(queueItems, 0, MediaStatus.REPEAT_MODE_REPEAT_OFF, null);
} catch (TransientNetworkDisconnectionException | NoConnectionException e) {
	e.printStackTrace();
}
```

I can confirm that indeed the app doesn't crash anymore when the video ends. But, now the user has to do additional step - tapping on the mini controller - in order to get to the Controller Activity.
How can I make the Controller Activity start right away?

**Comment by Nimrodda** [comment]
OK I managed to get the Cast Controller Activity to start. Please let me know if you see anything wrong with that. Full code:

In the `onClick` event I do:

```
MediaMetadata mediaMetadata = new MediaMetadata();
MediaInfo mediaInfo = new MediaInfo.Builder(URL)
			.setContentType("video/mp4")
			.setStreamType(MediaInfo.STREAM_TYPE_BUFFERED)
			.setMetadata(mediaMetadata).build();

MediaQueueItem mediaQueueItem = new MediaQueueItem.Builder(mediaInfo).setAutoplay(true).setPreloadTime(20).build();
MediaQueueItem[] queueItems = new MediaQueueItem[] { mediaQueueItem };
try {
		mCastManager.queueLoad(queueItems, 0, MediaStatus.REPEAT_MODE_REPEAT_OFF, null);
} catch (TransientNetworkDisconnectionException | NoConnectionException e) {
	e.printStackTrace();
}
mCastManager.addVideoCastConsumer(new OnQueueLoadedListener(getApplicationContext(), queueItems[0].getMedia()));
```

`OnQueueLoadedListener` implementation: 
```
private static class OnQueueLoadedListener extends VideoCastConsumerImpl {
    private Context mContext;
    private MediaInfo mMediaInfo;

public OnQueueLoadedListener(Context context, MediaInfo mMediaInfo) {
	this.mContext = context;
	this.mMediaInfo = mMediaInfo;
}

@Override
public void onMediaQueueUpdated(List<MediaQueueItem> queueItems, MediaQueueItem item, int repeatMode, boolean shuffle) {
	VideoCastManager castManager = VideoCastManager.getInstance();
	Intent intent = new Intent(mContext, castManager.getTargetActivity());
	intent.putExtra(VideoCastManager.EXTRA_MEDIA, Utils.mediaInfoToBundle(mMediaInfo));
	intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
	mContext.startActivity(intent);
	castManager.removeVideoCastConsumer(this);
}
}
```

**Comment by naddaf** [comment]
I'll take a look at your sample project and will get that working but it
has to wait till late night tonight to get to it.

**Comment by Nimrodda** [comment]
Alright, thanks a lot!

**Comment by Nimrodda** [comment]
Just updating you that the app no longer crashes since I move to the approach you recommended with a queue that has single item. :+1: My solution is basically identical to the code I pasted earlier.

**Comment by naddaf** [comment]
I still like to make sure that for those who don't use queue apis, things
still work without an issue :-)

**Comment by naddaf** [comment]
Adding the line
mMediaStatus = mCastManager.getMediaStatus();
to the beginning of the VideoCastControllerFragment#updatePlayerStatus() should fix this issue.

**Comment by Nimrodda** [comment]
OK cool! Simple solution :)

**Comment by naddaf** [comment]
This is addressed in 2.4
